mesonbuild/build.py

```diff
@@ -3003,6 +3004,143 @@ def get_classpath_args(self):
+            else:
+                bundle_layout = 'contents'
+
+        self.bundle_layout = bundle_layout
+
+        exe_dir_name = kwargs['bundle_exe_dir_name']
+
+        if exe_dir_name is None:
+            exe_dir_name = AppBundle._exe_dir_names_by_system.get(m.system, '')
+
+        self.exe_dir_name = exe_dir_name
+
+        if m.is_darwin() and m.subsystem != 'macos':
+            self.default_principal_class = 'UIApplication'
+        else:
+            self.default_principal_class = 'NSApplication'
+
+        info_plist = kwargs['info_plist']
+
+        if isinstance(info_plist, str):
+            if os.path.isabs(info_plist):
+                info_plist = File.from_absolute_file(info_plist)
+            else:
+                info_plist = File.from_source_file(self.environment.source_dir, self.subdir, info_plist)
+        elif isinstance(info_plist, File):
+            # When passing a generated file.
+            pass
+        elif isinstance(info_plist, (CustomTarget, CustomTargetIndex)):
+            # When passing output of a Custom Target
+            info_plist = File.from_built_file(info_plist.get_subdir(), info_plist.get_filename())
+        else:
+            raise InvalidArguments(
+                'info_plist must be either a string, '
+                'a file object, a Custom Target, or a Custom Target Index')
+
+        self.info_plist: T.Optional[File] = info_plist
+
+        # Names for these slightly named after their CoreFoundation names.
+        # 'oldstyle' is bundles with the executable in the top directory and a
+        # Resources folder, as used in Framework bundles and GNUstep,
+        # 'contents' is the standard macOS # application bundle layout with
+        # everything under a Contents directory, 'flat' is the iOS-style bundle
+        # with every file in the root directory of the bundle.
+        if self.bundle_layout not in ['oldstyle', 'contents', 'flat']:
+            raise InvalidArguments('bundle_layout must be one of \'oldstyle\', \'contents\', or \'flat\'.')
+
+    def type_suffix(self):
+        return "@app"
+
+    def contents_root(self) -> str:
+        return {
+            'oldstyle': [],
+            'contents': ['Contents'],
+            'flat': [],
+        }[self.bundle_layout]
+
+    def bin_root(self) -> T.List[str]:
+        return {
+            'oldstyle': [],
+            'contents': [x for x in ['Contents', self.exe_dir_name] if x],
+            'flat': [],
+        }[self.bundle_layout]
+
+    def resources_root(self) -> T.List[str]:
+        return {
+            'oldstyle': ['Resources'],
+            'contents': ['Contents', 'Resources'],
+            'flat': [],
+        }[self.bundle_layout]
+
+    def info_root(self) -> T.List[str]:
+        return {
+            # The only case where Info.plist is in the Resources directory.
+            'oldstyle': ['Resources'],
+            'contents': ['Contents'],
+            'flat': [],
+        }[self.bundle_layout]
```

Comment:
Surely there must be a better way to do this?

Reply:
Yeah, this is super ugly. I do want to find a better way to write this. Could at least make this return a str and keep the rest the same but I'm still not satisfied with that.

Reply:
There, I suppose the simplest way is probably the best here.